sdk/src/types/block/address/restricted.rs

```diff
@@ -0,0 +1,323 @@
+use getset::Getters;
+use packable::{error::UnpackErrorExt, prefix::BoxedSlicePrefix, Packable, PackableExt};
+
+use super::Address;
+use crate::types::block::Error;
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Getters, Packable)]
+#[getset(get = "pub")]
+pub struct RestrictedAddress {
+    address: Address,
+    allowed_capabilities: AddressCapabilities,
+}
+
+impl RestrictedAddress {
+    /// The [`Address`](crate::types::block::address::Address) kind of a [`RestrictedAddress`].
+    pub const KIND: u8 = 40;
+
+    /// Creates a new [`RestrictedAddress`] address from an [`Address`] with default allowed capabilities.
+    #[inline(always)]
+    pub fn new(address: impl Into<Address> + Send) -> Result<Self, Error> {
+        let address = address.into();
+        if matches!(address, Address::Restricted(_)) {
+            return Err(Error::InvalidAddressKind(Self::KIND));
+        }
+        Ok(Self {
+            address,
+            allowed_capabilities: Default::default(),
+        })
+    }
+
+    /// Sets the allowed capabilities flags.
+    #[inline(always)]
+    pub fn with_allowed_capabilities(mut self, allowed_capabilities: impl Into<AddressCapabilities>) -> Self {
+        self.allowed_capabilities = allowed_capabilities.into();
+        self
+    }
+
+    /// Sets the allowed capabilities flags.
+    #[inline(always)]
+    pub fn set_allowed_capabilities(&mut self, allowed_capabilities: impl Into<AddressCapabilities>) -> &mut Self {
+        self.allowed_capabilities = allowed_capabilities.into();
+        self
+    }
+}
+
+impl TryFrom<Address> for RestrictedAddress {
+    type Error = Error;
+
+    fn try_from(value: Address) -> Result<Self, Self::Error> {
+        Self::new(value)
+    }
+}
+
+impl core::fmt::Display for RestrictedAddress {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{}", prefix_hex::encode(self.pack_to_vec()))
+    }
+}
+
+/// All possible capabilities that an [`Address`] can have.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+#[non_exhaustive]
+pub enum AddressCapabilityFlag {
+    /// Can receive Outputs with Native Tokens.
+    NativeTokens,
+    /// Can receive Outputs with Mana.
+    Mana,
+    /// Can receive Outputs with a Timelock Unlock Condition.
+    OutputsWithTimelock,
+    /// Can receive Outputs with an Expiration Unlock Condition.
+    OutputsWithExpiration,
+    /// Can receive Outputs with a Storage Deposit Return Unlock Condition.
+    OutputsWithStorageDeposit,
+    /// Can receive Account Outputs.
+    AccountOutputs,
+    /// Can receive NFT Outputs.
+    NftOutputs,
+    /// Can receive Delegation Outputs.
+    DelegationOutputs,
+}
+
+impl AddressCapabilityFlag {
+    const NATIVE_TOKENS: u8 = 0b00000001;
+    const MANA: u8 = 0b00000010;
+    const OUTPUTS_WITH_TIMELOCK: u8 = 0b00000100;
+    const OUTPUTS_WITH_EXPIRATION: u8 = 0b00001000;
+    const OUTPUTS_WITH_STORAGE_DEPOSIT: u8 = 0b00010000;
```

Comment:
`OUTPUTS_WITH_STORAGE_DEPOSIT_RETURN`?

Reply:
yes this is so much better now I can ride the variable names all the way to the moon /s

Reply:
`storage_score_offset_implicit_account_creation_address` enters the chat ...

Reply:
I agree on this, all outputs have storage deposit, we are really speaking of the return here